nlp/model/augmented_lagrangian.py

```diff
@@ -128,6 +129,21 @@ def hess(self, *args, **kwargs):
         return self.hop(*args, **kwargs)
 
 
+    def magical_step(self, x, g):
+        """Compute a magical step for the slack variables.
+
+        This step minimizes the augmented Lagrangian with respect to the slack
+        variables only for a fixed set of decision variables. Here, g is the 
+        gradient computed by self.grad at x.
+        """
+        model = self.model
+        on = model.original_n
+        m_step = np.zeros(self.n)
+        m_step[on:] = -g[on:] / self.penalty
+        m_step = projected_step(x, m_step, model.Lvar, model.Uvar)
+        return m_step
+
+
```

Comment:
Do you want me to remove the "magical steps" changes in this PR so you can close it? I can finish the magical steps in a different branch / pull request since there will be more work to do.